### PR TITLE
IGNITE-17959 Added retrying in ReplicaService for a case when an operation throws ReplicaUnavailableException in ReplicaManager.

### PR DESCRIPTION
**Pull request by sergeyuttsel**
https://issues.apache.org/jira/browse/IGNITE-17959

### CODE REVIEW

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 1]
```diff
@@ -82,6 +85,9 @@ public class ReplicaManager implements IgniteComponent {
     /** Replicas. */
     private final ConcurrentHashMap<ReplicationGroupId, Replica> replicas = new ConcurrentHashMap<>();
 
+    /** Replicas which was requested but have not started yet. */
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:88]: that were

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/message/AwaitReplicaRequest.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+package org.apache.ignite.internal.replicator.message;
+
+import org.apache.ignite.internal.replicator.Replica;
+import org.apache.ignite.internal.replicator.ReplicaManager;
+import org.apache.ignite.internal.replicator.ReplicaService;
+import org.apache.ignite.internal.replicator.exception.ReplicaUnavailableException;
+import org.apache.ignite.network.annotations.Transferable;
+
+/**
+ * This request is sent when some request fails in {@link ReplicaService} with an {@link ReplicaUnavailableException}
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/message/AwaitReplicaRequest.java:27]: Please rephrase.

**Review by sanpwc — CHANGES_REQUESTED** [review]

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (sergeyuttsel) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java]: Seems that it isn't actual now.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 3]
```diff
@@ -82,6 +85,9 @@ public class ReplicaManager implements IgniteComponent {
...
+    private final Map<ReplicationGroupId, CompletableFuture<Void>> pendingReplicas = new ConcurrentHashMap<>();
```
[the remark this answers is not in the document]
Comment (sergeyuttsel) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java]: I removed the pendingReplicas.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/message/AwaitReplicaRequest.java (hunk 2 above)
Comment (sergeyuttsel) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/message/AwaitReplicaRequest.java]: Fixed. Check again please.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/message/AwaitReplicaResponse.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+package org.apache.ignite.internal.replicator.message;
+
+import org.apache.ignite.network.annotations.Transferable;
+
+/**
+ * Await replica response.
+ */
+@Transferable(ReplicaMessageGroup.AWAIT_REPLICA_RESPONSE)
+public interface AwaitReplicaResponse extends ReplicaResponse {
```
[the remark this answers is not in the document]
Comment (sergeyuttsel) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/message/AwaitReplicaResponse.java:26]: As we decided I didn't fix it.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 5]
```diff
@@ -205,15 +232,19 @@ private Replica startReplicaInternal(
...
 
-        if (previous != null) {
-            throw new ReplicaIsAlreadyStartedException(replicaGrpId);
-        }
+                return replicaFut;
+            }
+        });
 
-        return replica;
+        return replicas.get(replicaGrpId).join();
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:247]: Let's add a comment that replicaFut is always completed here.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 6]
```diff
@@ -330,6 +361,21 @@ private void sendReplicaUnavailableErrorResponse(
         }
     }
 
+    /**
+     * Sends await replica response.
+     */
+    private void sendAwaitReplicaResponse(
+            ClusterNode sender,
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:368]: two params should be in one line.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 7]
```diff
@@ -372,10 +418,10 @@ private NetworkMessage prepareReplicaErrorResponse(HybridTimestamp requestTimest
     private void idleSafeTimeSync() {
         replicas.values().forEach(r -> {
             ReplicaSafeTimeSyncRequest req = REPLICA_MESSAGES_FACTORY.replicaSafeTimeSyncRequest()
-                    .groupId(r.groupId())
+                    .groupId(r.join().groupId())
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:421]: Let's skip uncompleted replicas there's no sense in sending safeTimeSyncRequests to them.

**Review by sanpwc — CHANGES_REQUESTED** [review]

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaService.java [hunk 8]
```diff
@@ -100,14 +112,57 @@ private <R> CompletableFuture<R> sendToReplica(ClusterNode node, ReplicaRequest
 
                 if (response instanceof ErrorReplicaResponse) {
                     var errResp = (ErrorReplicaResponse) response;
-                    res.completeExceptionally(errResp.throwable());
+
+                    if (errResp.throwable() instanceof ReplicaUnavailableException) {
+                        pendingInvokes.compute(node, (clusterNode, fut) -> {
+                            AwaitReplicaRequest awaitReplicaReq = REPLICA_MESSAGES_FACTORY.awaitReplicaRequest()
```
Comment (sanpwc) [inline modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaService.java:118]: Seems that we may put message preparation inside fut == null clause.

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 9]
```diff
@@ -118,17 +120,42 @@ public ReplicaManager(
...
-                HybridTimestamp requestTimestamp = extractTimestamp(request);
+                if (request instanceof AwaitReplicaRequest) {
+                    replicas.compute(request.groupId(), (replicationGroupId, replicaFut) -> {
+                        if (replicaFut == null) {
+                            replicaFut = new CompletableFuture<>();
+                        }
+
+                        if (!replicaFut.isDone()) {
+                            replicaFut.thenCompose(ignore -> {
+                                IgniteUtils.inBusyLock(busyLock,
```
[the remark this answers is not in the document]
Comment (sergeyuttsel) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java]: fixed

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 10]
```diff
@@ -118,17 +120,42 @@ public ReplicaManager(
 
                 ReplicaRequest request = (ReplicaRequest) message;
 
-                HybridTimestamp requestTimestamp = extractTimestamp(request);
+                if (request instanceof AwaitReplicaRequest) {
```
[the remark this answers is not in the document]
Comment (sergeyuttsel) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:124]: fixed

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 11]
```diff
@@ -118,17 +120,42 @@ public ReplicaManager(
...
 
-                if (replica == null) {
+                if (replicaFut == null || !replicaFut.isDone()) {
                     sendReplicaUnavailableErrorResponse(sender, correlationId, request, requestTimestamp);
 
                     return;
                 }
 
-                CompletableFuture<Object> result = replica.processRequest(request);
+                CompletableFuture<Object> result = replicaFut.join().processRequest(request);
```
[the remark this answers is not in the document]
Comment (sergeyuttsel) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:164]: fixed

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java [hunk 12]
```diff
@@ -164,7 +191,7 @@ public Replica replica(ReplicationGroupId replicaGrpId) throws NodeStoppingExcep
         }
 
         try {
-            return replicas.get(replicaGrpId);
+            return replicas.get(replicaGrpId).join();
```
[the remark this answers is not in the document]
Comment (sergeyuttsel) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java]: fixed

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 5 above)
Comment (sergeyuttsel) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java:235]: fixed

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 6 above)
Comment (sergeyuttsel) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java]: fixed

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java (hunk 7 above)
Comment (sergeyuttsel) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java]: fixed

File: modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaService.java (hunk 8 above)
Comment (sergeyuttsel) [reply modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaService.java]: fixed

**Review by sanpwc — APPROVED** [review]

**Pull request merged**